pkg/driver/controller_server.go

```diff
@@ -1350,28 +1369,45 @@ func (driver *Driver) ListSnapshots(ctx context.Context, request *csi.ListSnapsh
 	// request.StartingToken <- use me
 	// request.MaxEntries <- use me
 
-	// If driver doesn't know any storage providers yet, then return empty list with success
-	if len(driver.storageProviders) == 0 {
-		log.Info("No storage providers are known to the CSI driver yet.")
-		return &csi.ListSnapshotsResponse{
-			Entries: []*csi.ListSnapshotsResponse_Entry{},
-		}, nil
-	}
-
-	// TODO: ListSnapshots does not provide Secrets nor a NodeId so it simply needs to return information about all of the
-	// snapshots it knows about.  This can be implemented by reading all snapshots from every array known to the driver.  We can do this
-	// at initialization time and keep them in memory.
 	var allSnapshots []*model.Snapshot
-	for _, storageProvider := range driver.storageProviders {
-		// TODO: skip storage providers that are not of the same vendor... or not.  Need to think through this
-		snapshots, err := storageProvider.GetSnapshots(request.SourceVolumeId)
+
+	// Secrets are optional
+	if request.Secrets != nil {
+		storageProvider, err := driver.GetStorageProvider(request.Secrets)
 		if err != nil {
-			log.Trace("err: ", err.Error())
-			return nil, status.Error(codes.Internal, "Error while attempting to list snapshots")
+			log.Error("err: ", err.Error())
+			return nil, status.Error(codes.Unavailable, fmt.Sprintf("Failed to get storage provider from secrets, %s", err.Error()))
 		}
-		log.Tracef("Read %d snapshots from a storage provider", len(snapshots))
-		for _, snapshot := range snapshots {
-			allSnapshots = append(allSnapshots, snapshot)
+		allSnapshots, err = getSnapshotsForStorageProvider(ctx, request, storageProvider)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// In the event that ListSnapshots does not provide Secrets it simply needs to return information about all of the
+		// snapshots it knows about.  This can be implemented by reading all snapshots from every array known to the driver.  We can do this
+		// at initialization time and keep them in memory.
+
+		// If driver doesn't know any storage providers yet, then return empty list with success
+		if len(driver.storageProviders) == 0 {
+			log.Info("No storage providers are known to the CSI driver yet.")
+			return &csi.ListSnapshotsResponse{
+				Entries: []*csi.ListSnapshotsResponse_Entry{},
+			}, nil
+		}
+
+		for _, storageProvider := range driver.storageProviders {
+			snapshots, err := getSnapshotsForStorageProvider(ctx, request, storageProvider)
+			if err != nil {
```

Comment:
we should not error out here then as it might be wrong storageprovide for that volume/snapshotid.

Reply:
I don't think we should ignore all errors here.  If it is the wrong storage provider, we will get a 404 error and err will not be nil.  If there is an auth issue, we get a 401 and login again.

If it is the right storage provider but there's an error, we need to log it and return.

Reply:
yes, this will be good. we seem to get error codes in string, so can lookup these.